### PR TITLE
[FEATURE] Add Semi-SL multilabel classification algorithm

### PR DESCRIPTION
**Pull request by sovrasov**
- SemiSL for multilabel classification with BT loss and adaptive loss weighting
- Docs
- CLI test
- Unit test

Future work:
- Add SemiSL to regression tests

### CODE REVIEW

**Review by sungmanc — DISMISSED** [review]
LGTM

File: otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        if self.use_dynamic_loss_weighting:
+            losses["loss"] = self.loss_balancer.balance_losses((l_labeled, aux_loss))
+        else:
+            losses["loss"] = l_labeled + self.unlabeled_coef * aux_loss
+        losses["unlabeled_loss"] = self.unlabeled_coef * aux_loss
+
+        return losses
+
+    def forward_train_with_last_layers(self, x, gt_label, final_cls_layer, final_emb_layer):
+        """Forwards multilabel semi-ls head and losses
```
Comment (JihwanEom) [inline otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py]:
```suggestion
        """Forwards multilabel semi-sl head and losses
```

File: otx/cli/manager/config_manager.py [hunk 2]
```diff
@@ -437,6 +437,7 @@ def build_workspace(self, new_workspace_path: Optional[str] = None) -> None:
         # Copy config files
         config_files = [
             (model_dir, "model.py", train_type_dir),
+            (model_dir, "model_multilabel.py", train_type_dir),
```
Comment (harimkang) [inline otx/cli/manager/config_manager.py:440]: We modified it to import `model_multilabel.py` based on model dir in this line, then we can delete line 446 `(template_dir, "model_multilabel.py", self.workspace_root),` right?

Comment (sovrasov) [reply otx/cli/manager/config_manager.py:440]: I guess yes, you're right

File: otx/mpa/modules/models/classifiers/semisl_multilabel_classifier.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+class SemiSLMultilabelClassifier(SAMImageClassifier):
+    """Semi-SL Multilabel Classifier
+    This classifier supports unlabeled data by overriding forward_train
+    """
+
+    def forward_train(self, imgs, gt_label, **kwargs):
+        """Data is transmitted as a classifier training function
+
+        Args:
+            imgs (list[Tensor]): List of tensors of shape (1, C, H, W)
```
Comment (kprokofi) [inline otx/mpa/modules/models/classifiers/semisl_multilabel_classifier.py:24]: Can you add gt_label to doc string as well?

File: otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        return self.forward_train_with_last_layers(
+            x, gt_label, final_cls_layer=self.classifier, final_emb_layer=self.aux_mlp
+        )
+
+
+def generate_aux_mlp(aux_mlp_cfg: dict, in_channels: int):
+    out_channels = aux_mlp_cfg["out_channels"]
+    if out_channels <= 0:
+        raise ValueError(f"out_channels={out_channels} must be a positive integer")
+    if "hid_channels" in aux_mlp_cfg and aux_mlp_cfg["hid_channels"] > 0:
```
Comment (kprokofi) [inline otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py]: Maybe we can fix configuration for the `aux_mlp_cfg` (and other hyperparameters like that)  and remove if statement. 
I just want to remove additional complexity. We can fix some additional hyperparameters that we unlikely will change in the future and remove the variability of the algorithm at this level.

File: otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        logits = final_cls_layer(x["labeled_weak"])
+        features_weak = torch.cat((final_emb_layer(x["labeled_weak"]), final_emb_layer(x["unlabeled_weak"])))
+        features_strong = torch.cat((final_emb_layer(x["labeled_strong"]), final_emb_layer(x["unlabeled_strong"])))
+        features = (features_weak, features_strong)
+        losses = self.loss(logits, gt_label, features)
+        return losses
+
+
+@HEADS.register_module()
+class SemiLinearMultilabelClsHead(SemiMultilabelClsHead, CustomMultiLabelLinearClsHead):
```
Comment (kprokofi) [inline otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py:94]: Maybe we can merge these classes during refactoring. I'm not sure, but now we have different heads for MNV3 and EffNet - it is not convenient and starts to spread out

File: otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        self.reset()
+
+    def reset(self):
+        self.val = 0
+
+    def update(self, val):
+        self.val = self.alpha * self.val + (1 - self.alpha) * val
+
+
+class LossBalancer:
```
Comment (kprokofi) [inline otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py]: Maybe move these classes to some utils folder?

File: docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst [hunk 7]
```diff
@@ -52,11 +52,47 @@ In the table below the `mAP <https://en.wikipedia.org/w/index.php?title=Informat
...
+|                       |   SL                 | Semi-SL |  SL            | Semi-SL |   SL           | Semi-SL |
++-----------------------+----------------------+---------+----------------+---------+----------------+---------+
+| MobileNet-V3-large-1x |  74.28               |  74.41  | 96.34          |  97.29  |  82.39         |  83.77  |
++-----------------------+----------------------+---------+----------------+---------+----------------+---------+
+|   EfficientNet-B0     |  79.59               |  80.91  | 97.75          |  98.59  | 83.24          |  84.19  |
++-----------------------+----------------------+---------+----------------+---------+----------------+---------+
+|  EfficientNet-V2-S    |  75.91               |  81.91  | 95.65          |  96.43  | 85.19          |  84.24  |
++-----------------------+----------------------+---------+----------------+---------+----------------+---------+
+
+AerialMaritime was sampled with 5 images per label. VOC and COCO were sampled with 40 images per label.
```
Comment (kprokofi) [inline docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst]:
```suggestion
AerialMaritime was sampled with 5 images per category. VOC and COCO were sampled with 40 images per category.
```
or class

File: docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst [hunk 8]
```diff
@@ -52,11 +52,47 @@ In the table below the `mAP <https://en.wikipedia.org/w/index.php?title=Informat
...
-.. Semi-supervised Learning
-.. ************************
+************************
+Semi-supervised Learning
+************************
 
-.. To be added soon
+Semi-SL (Semi-supervised Learning) is a type of machine learning algorithm that uses both labeled and unlabeled data to improve the performance of the model. This is particularly useful when labeled data is limited, expensive or time-consuming to obtain.
+
+To utilize unlabeled data during training, we use `BarlowTwins loss <https://arxiv.org/abs/2103.03230>`_ as an auxiliary loss for Semi-SL task solving. BarlowTwins enforces consistency across augmented versions of the same data (both labeled and unlabeled). Additionally, our algorithm uses a combination of strong data augmentations and a specific optimizer called Sharpness-Aware Minimization (SAM) to further improve the accuracy of the model.
```
Comment (kprokofi) [inline docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst]:
```suggestion
To utilize unlabeled data during training multilabel classification model, we use `BarlowTwins loss <https://arxiv.org/abs/2103.03230>`_ as an auxiliary loss for Semi-SL  task solving. BarlowTwins enforces consistency across augmented versions of the same data (both labeled and unlabeled). Additionally, our algorithm uses a combination of strong data augmentations and a specific optimizer called Sharpness-Aware Minimization (SAM) to further improve the accuracy of the model.
```

Comment (kprokofi) [reply docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst]: Please add link for SAM

File: docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst [hunk 9]
```diff
@@ -52,11 +52,47 @@ In the table below the `mAP <https://en.wikipedia.org/w/index.php?title=Informat
...
+
+Overall, OpenVINO™ Training Extensions utilizes powerful techniques for improving the performance of Semi-SL algorithm with limited labeled data. They can be particularly useful in domains where labeled data is expensive or difficult to obtain, and can help to reduce the time and cost associated with collecting labeled data.
```
Comment (kprokofi) [inline docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst]: I think this sentence is the same as the first one and doesn't give  us additional info. Consider to remove it

File: docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst [hunk 10]
```diff
@@ -52,11 +52,47 @@ In the table below the `mAP <https://en.wikipedia.org/w/index.php?title=Informat
...
+
+.. _mlc_cls_semi_supervised_pipeline:
+
+- ``BarlowTwins loss``: A specific implementation of Semi-SL that combines the use of a consistency loss with strong data augmentations, and a specific optimizer called Sharpness-Aware Minimization (SAM) to improve the performance of the model.
+
+- ``Adaptive loss auxiliary loss weighting``: A technique for assigning such a weight for an auxiliary loss that the resulting value is a predefined fraction of the EMA-smoothed main loss value. This method allows aligning contribution of the losses during different training phases.
+
+- ``Exponential Moving Average (EMA)``: A technique for maintaining a moving average of the model's parameters, which can improve the generalization performance of the model.
+
+- ``Additional techniques``: Other than that, we use several solutions that apply to supervised learning (No bias Decay, Augmentations, Early-Stopping, etc.)
```
Comment (kprokofi) [inline docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst:71]: You also use "strong" and "weak" pipelines, could you describe them here?

File: docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst [hunk 11]
```diff
@@ -52,11 +52,47 @@ In the table below the `mAP <https://en.wikipedia.org/w/index.php?title=Informat
...
+
+Please, refer to the :doc:`tutorial <../../../tutorials/advanced/semi_sl>` on how to train semi-supervised learning.
+Training time depends on the number of images and can be up to several times longer than conventional supervised learning.
+
+In the table below the mAP on some academic datasets using our pipeline is presented.
```
Comment (kprokofi) [inline docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst]:
```suggestion
In the table below the mAP on some academic datasets using our pipeline is presented.
```
```suggestion
In the table below the mAP metric on some public datasets using our pipeline is presented.
```
I am not sure that Aerial Maritime is popular academic benchmark dataset

Comment (kprokofi) [reply docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst]: Also, please, describe more in depth how many images you utilized for this experiments

File: docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst [hunk 12]
```diff
@@ -52,11 +52,47 @@ In the table below the `mAP <https://en.wikipedia.org/w/index.php?title=Informat
...
++-----------------------+----------------------+---------+----------------+---------+----------------+---------+
+|   EfficientNet-B0     |  79.59               |  80.91  | 97.75          |  98.59  | 83.24          |  84.19  |
++-----------------------+----------------------+---------+----------------+---------+----------------+---------+
+|  EfficientNet-V2-S    |  75.91               |  81.91  | 95.65          |  96.43  | 85.19          |  84.24  |
++-----------------------+----------------------+---------+----------------+---------+----------------+---------+
+
+AerialMaritime was sampled with 5 images per label. VOC and COCO were sampled with 40 images per label.
+
+.. note::
+    This result can vary greatly depending on the image selected for each class. Also, since there are few labeled settings for the Semi-SL algorithm. Some models may require larger datasets for better results.
```
Comment (kprokofi) [inline docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst]:
```suggestion
    This result greatly depends on the image selected for each class. Also, since there are few labeled settings for the Semi-SL algorithm. Some models may require larger datasets for better results.
```

File: otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py (hunk 5 above)
Comment (sovrasov) [reply otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py:94]: We can add a task to the refactoring epic

File: otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py (hunk 4 above)
Comment (sovrasov) [reply otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py]: Users barely will touch the configs, but from the development perspective it is convenient to have an advanced configuration.

File: docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst (hunk 10 above)
Comment (sovrasov) [reply docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst:71]: I described them in general description to clarify work of the BT loss.

File: docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst (hunk 9 above)
Comment (sovrasov) [reply docs/source/guide/explanation/algorithms/classification/multi_label_classification.rst]: Deleted

File: otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py (hunk 6 above)
Comment (sovrasov) [reply otx/mpa/modules/models/heads/semisl_multilabel_cls_head.py]: Done

File: otx/mpa/modules/models/classifiers/semisl_multilabel_classifier.py (hunk 3 above)
Comment (sovrasov) [reply otx/mpa/modules/models/classifiers/semisl_multilabel_classifier.py:24]: Added

**Review by harimkang — APPROVED** [review]
I think this is great work!
The only thing I would say is that the E2E test is missing. Other than that, I have no comments.

**Review by JihwanEom — APPROVED** [review]

**Review by kprokofi — APPROVED** [review]
LGTM
